Add rsa sign to fips provider

slontis opened the pull request:
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

levitte commented on crypto/rsa/rsa_pmeth.c:
```diff
@@ -153,7 +154,9 @@ static int pkey_rsa_sign(EVP_PKEY_CTX *ctx, unsigned char *sig,
-        } else if (rctx->pad_mode == RSA_X931_PADDING) {
+        } else
+#endif /* FIPS_MODE */
+        if (rctx->pad_mode == RSA_X931_PADDING) {
```
Why are you changing this file at all?  It should *never* become part of the FIPS module, this is a legacy backend.

If there are things here that are missing in `providers/implementations/signature/rsa.c`, then they should be added there.

levitte commented on providers/implementations/signature/build.info:
```diff
@@ -2,12 +2,10 @@
-
-
+SOURCE[../../libfips.a]=rsa.c
+SOURCE[../../libnonfips.a]=rsa.c
```
Hmmmm, I know why, and it's not the cleanest...  some of the signing code could be placed in crypto/rsa/, especially considering how much is shared between crypto/rsa/rsa_pmeth.c and providers/implementations/signature/rsa.c...

I'm making a mental note on this for the moment being...

levitte commented on crypto/rsa/rsa_local.h:
```diff
@@ -48,12 +48,12 @@ struct rsa_st {
     BIGNUM *dmp1;
     BIGNUM *dmq1;
     BIGNUM *iqmp;
+    RSA_PSS_PARAMS *pss;
```
Why did you leave the comment behind?

slontis replied:
Oops..

levitte replied:
You misunderstood.  The comment on line 55 should be moved up.
(but yeah, removing the TODO was a good move nonetheless)

levitte replied:
You moved one comment too much.  The "be careful..." one belongs with `ex_data`

slontis replied:
Ok moved it to the correct line (which is not where it was to start with) :)

slontis commented:
Note the test for this is deferred to PR #11036.. (i.e running the self test for rsa sign/verify confirmed that this is working in the fips module).

slontis commented:
Taken out of draft mode now that I have confirmed that the sign/verify functions correctly..

levitte commented on providers/fips/fipsprov.c:
```diff
@@ -910,6 +912,9 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
+            break;
         case OSSL_FUNC_CRYPTO_MALLOC:
             c_CRYPTO_malloc = OSSL_get_CRYPTO_malloc(in);
             break;
```
Note for future cleanup: we should really consider having common provider functions for this sort of thing.  I tried to start that with `providers/common/include/prov/bio.h` and `providers/common/bio_prov.c`

levitte commented on crypto/provider_core.c:
```diff
@@ -893,8 +893,8 @@ static const OSSL_DISPATCH core_dispatch_[] = {
-    { OSSL_FUNC_CORE_POP_ERROR_TO_MARK,
-      (void (*)(void))core_pop_error_to_mark },
+    { OSSL_FUNC_CORE_POP_ERROR_TO_MARK, (void (*)(void))core_pop_error_to_mark },
+    { OSSL_FUNC_CORE_VSNPRINTF, (void (*)(void))BIO_vsnprintf },
```
Mind moving this line down a bit, so it's together with `OSSL_FUNC_BIO_VPRINTF`?

slontis replied:
No I dont mind - but it doesnt have too much to do with BIO's

levitte replied:
It's in the BIO API...  but otherwise, you're remark is quite right.  However, they're the same kind of function, so...
(this is like discussing that while fprintf is a FILE function, printf isn't, and yet, they are usually presented together)

levitte commented:
Rebase needed.

slontis commented:
rebased..

slontis commented:
> Also, if you're only going to have a direct reference to `BIO_vsnprintf`, the number macro should

Renamed and moved any related code around to be near the other bio stuff.

slontis commented:
@levitte - this should address your comments..

openssl-machine commented:
This pull request is ready to merge

slontis commented:
Thanks. Merged to master.
